code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,245 @@ public void updateAccount(Account account, String name, String kind, String exte
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
     }
     
-    public ObjectManager getObjectManager() {
-        return objectManager;
+    public boolean isProjectMember (long projectId, Long usingAccount, boolean isAdmin,Set<ExternalId> externalIds){
+        if (externalIds == null) {
+            return false;
+        }
+        if ((usingAccount != null && usingAccount.equals(projectId)) || isAdmin) {
+            return true;
+        }
+        Set<String> roles =  new HashSet<>();
+        roles.add(ProjectConstants.OWNER);
+        roles.add(ProjectConstants.MEMBER);
+        Set<ProjectMemberRecord> projectMembers = new HashSet<>();
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.ROLE.in(roles))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
     }
 
-    @Inject
-    public void setObjectManager(ObjectManager objectManager) {
-        this.objectManager = objectManager;
+    @Override
+    public Account updateProject(Account project, String name, String description) {
+        Map<TableField<AccountRecord, String>, String> properties = new HashMap<>();
+        if (name != null) {
+            properties.put(ACCOUNT.NAME, name);
+        }
+        if (description != null) {
+            properties.put(ACCOUNT.DESCRIPTION, description);
+        }
+        if (name == null && description == null){
+            return project;
+        }
+        int updateCount = create()
+                .update(ACCOUNT)
+                .set(properties)
+                .where(ACCOUNT.ID
+                        .eq(project.getId()))
+                .execute();
+        if(1 != updateCount) {
+            throw new RuntimeException("UpdateProject failed.");
+        }
+        return getAccountById(project.getId());
     }
 
-    public GenericResourceDao getGenericResourceDao() {
-        return resourceDao;
+    @Override
+    public List<? extends ProjectMember> setProjectMembers(final Account project, final HashSet<Member> members) {
+        return lockManager.lock(new ProjectLock(project), new LockCallback<List<? extends ProjectMember>>() {
+            public List<? extends ProjectMember> doWithLock() {
+                List<? extends ProjectMember> previousMembers = getActiveProjectMembers(project.getId());
```

Comment:
@alena1108 How would using executeStandardProcess create and update for the members be in your book? This would make this resource action synchronous and allow no out of sync states for setProjectMembers() to start with inaccurate list of members.